drivers/MbedCRC.h

```diff
@@ -0,0 +1,463 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_CRC_API_H
+#define MBED_CRC_API_H
+
+#include <stdint.h>
+#include "drivers/TableCRC.h"
+
+#if   defined ( __CC_ARM )
+#pragma diag_suppress 62  // Shift count is negative
+#elif defined ( __GNUC__ )
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wshift-count-negative"
+#endif
+
+namespace mbed {
+/** \addtogroup drivers */
+/** @{*/
+
+/** CRC Polynomial value
+ *
+ * Different polynomial values supported
+ */
+typedef enum crc_polynomial {
+    POLY_OTHER = 0,
+    POLY_8BIT_CCITT = 0x07,         // x8+x2+x+1
+    POLY_7BIT_SD = 0x9,             // x7+x3+1;
+    POLY_16BIT_CCITT = 0x1021,      // x16+x12+x5+1
+    POLY_16BIT_IBM = 0x8005,        // x16+x15+x2+1
+    POLY_32BIT_ANSI = 0x04C11DB7,    // x32+x26+x23+x22+x16+x12+x11+x10+x8+x7+x5+x4+x2+x+1
+} crc_polynomial_t;
+
+/** CRC object provides CRC generation through hardware/software
+ *
+ *  @tparam  polynomial CRC polynomial value in hex
+ *  @tparam  width CRC polynomial width
+ */
+
+template <uint32_t polynomial=POLY_32BIT_ANSI, uint8_t width=32>
+class MbedCRC
+{
+public:
+    typedef uint64_t crc_data_size_t;
+
+    /** Lifetime of CRC object
+     *
+     *  @param  initial_xor  Inital value/seed to Xor (Default ~0x0)
+     *  @param  final_xor  Final Xor value (Default ~0x0)
+     *  @param  reflect_data
+     *  @param  reflect_remainder
+     */
+    MbedCRC(uint32_t initial_xor, uint32_t final_xor, bool reflect_data, bool reflect_remainder);
+    MbedCRC();
+    virtual ~MbedCRC()
+    {
+        // Do nothing
+    }
+
+    /** Initialize a CRC module
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    int32_t init(void)
```

Comment:
I suppose this will do something when hardware CRC will be available. Is it a correct assumption ?

Reply:
Yes.

Reply:
I'm not sure about that, it's an interface, I wouldn't like people to use it:
```
MbedCRC crc(...);
crc.init();
crc.compute(...);
crc.deinit();
```

Can't we just use constructor/destructor for (de)init code?